broadinstitute/single_cell_portal_core
# Base64 encode custom color annotation/label names to avoid MongoDB (SCP-4151)

This update addresses an issue in our version of MongoDB where BSON documents cannot have key names with either periods or dollar signs (`.`, `$`).  This constraint is due to both being control characters that have special function (periods are used for hash key lookup notation, and dollar signs denote clauses/actions in queries).  This is not an issue in other parts of the application where we control the keys being saved, but for custom color objects in `ClusterFileInfo`, the key names map to annotation and label names, all of which are user-generated.

To get around this issue, this update now base64 encodes hash key names for `ClusterFileInfo.custom_colors` documents.  A helper method called `transform_custom_colors` is used to both encode and decode nested hashes, and a convenience method of `decoded_custom_colors` can be used in the UI for automatically decoding values at runtime.  In addition, a reversible database migration has been added to covert existing data in the database.

MANUAL TESTING
1. Boot as normal and sign in
2. In a separate terminal window, ensure caching is turned off with `bin/rails dev:cache` and look for the message `Development mode is no longer being cached.`
3. Load any study with existing clusters, and update the colors for one annotation and save
4. Upload the file `test_data/validation/cluster_example_with_periods.txt` to this same study
5. Back in the other terminal window, run the database migration with `bin/rails db:migrate` (doing this after saving a custom color object will validate the migration transform works)
6. Reload the study, and validate the custom colors you saved earlier still work
7. Load the `cluster_example_with_periods.txt` cluster, and set custom colors for the `annotation.with.periods` annotation
8. Confirm that colors save and render correctly

This PR satisfies SCP-4151.
